### ISSUE TITLE
HardFault when running USB Mass Storage CDROM example

### ISSUE THREAD

**Issue by audiofish** [OpenNuvoton/NUC121_125BSP#5]
I have been trying to run the "USBD_MassStorage_CDROM" sample code.

However, I am continuously running into issues with the code reaching a HardFault.

Typically this results in a fault with the following stack trace:

> 
> Breakpoint 1, HardFault_Handler () at NUC121_125BSP/Library/StdDriver/src/retarget.c:394
> 394     }
> (gdb) bt
> #0  HardFault_Handler () at NUC121_125BSP/Library/StdDriver/src/retarget.c:394
> #1  signal handler called
> #2  USBD_MemCopy (size=55, src=0x20002001 "p\002", dest=0x40060151 "") at NUC121_125BSP/Library/StdDriver/inc/usbd.h:593
> #3  MSC_Read () at NUC121_125BSP/SampleCode/StdDriver/USBD_MassStorage_CDROM/MassStorage.c:788
> #4  0x00000b46 in MSC_AckCmd () at NUC121_125BSP/SampleCode/StdDriver/USBD_MassStorage_CDROM/MassStorage.c:1653
> #5  0x00000bec in EP2_Handler () at NUC121_125BSP/SampleCode/StdDriver/USBD_MassStorage_CDROM/MassStorage.c:388
> #6  0x00000d04 in USBD_IRQHandler () at NUC121_125BSP/SampleCode/StdDriver/USBD_MassStorage_CDROM/MassStorage.c:346
> #7  signal handler called
> #8  0x0000171a in main () at NUC121_125BSP/SampleCode/StdDriver/USBD_MassStorage_CDROM/main.c:159
>

**Comment by egbert-h** [comment]
Hi  audiofish,

I also tried it with latest version and seems it work in WIN10.

Could you provide the information of compiling environment 

and the host OS?

thanks .

**Comment by audiofish** [comment]
Hi egbert-h,

I believe I have tracked it down to an incompatibility with a Linux host OS. If I run the example under Windows 10 I get the correct result. However, trying it under Ubuntu 20.10 I see the fault. Code was compiled within Keil uVision 5.

I captured some comparative USB traces with Wireshark under Windows and Linux, attached to this comment. (I have zipped them because GitHub does not support this file type).
[usb_captures.zip](https://github.com/OpenNuvoton/NUC121_125BSP/files/7017722/usb_captures.zip)

Both of the traces look correct until packet number 122 on the Windows trace and packet number 163, where the host sends a "Get Configuration" message. When the device responds with a "Get Configuration Response Data" packet, it looks malformed in both captures:
The USB Mass Storage packet does not get decoded, presumably because a "USBC" signature is sent, rather than a "USBS" signature that is expected for a response. This gets incorrectly decoded as the SCSI Payload data length instead, giving an obviously bogus length of 1431519811 bytes. The packet is truncated, I think because the host has only allocated 8 bytes for the response (although I'm less sure about the details of the protocol here). The device then also sends a "Get Configuration Response Data" with a status "Good".

The Windows driver appears to accept this malformed packet okay and continue. However, Linux driver re-sends the "Get Configuration" message, with a much larger allocation of 65530 bytes. I think this is where the device code breaks. The "Get Configuration Response Data" message returned is 7808 bytes of malformed data and then stops responding (presumably because this is when it HardFaults).

I'm having a bit of trouble following the control flow of the code in MassStorage.c to debug this.
